charts/tfy-karpenter-config/files/al2023-soci-default-provisioner-userdata.sh

```diff
@@ -1,133 +1,63 @@
+        [proxy_plugins.soci.exports]
+          root = "/var/lib/soci-snapshotter-grpc"
+      [plugins."io.containerd.grpc.v1.cri".containerd]
+        snapshotter = "soci"
+        # This line is required for containerd to send information about how to lazily load the image to the snapshotter
+        disable_snapshot_annotations = false
+
+--BOUNDARY
+Content-Type: text/x-shellscript; charset="us-ascii"
+
 #!/bin/bash
 set -ex
-
-version_lte() {
-  printf '%s\n' "$1" "$2" | sort -C -V
-}
-
-version_lt() {
-  ! version_lte "$2" "$1"
-}
-
-CONTAINERD_VERSION_OUTPUT="$(containerd --version)"
-IFS=" " read -a  CONTAINERD_VERSION_STRING <<< "${CONTAINERD_VERSION_OUTPUT}"
-CONTAINERD_VERSION=${CONTAINERD_VERSION_STRING[2]}
-SANDBOX_IMAGE=$(containerd config dump | grep -oE 'sandbox_image = "(.*)"' | sed 's/sandbox_image = //')
-KUBELET_CONFIG_DIR="/etc/kubernetes/kubelet/config.json.d"
-SOCI_KUBELET_CONFIG_PATH="$KUBELET_CONFIG_DIR/99-soci.conf"
-CONTAINERD_CONFIG_FILEPATH="/etc/containerd/config.toml"
-BACKUP_CONTAINERD_CONFIG_FILEPATH="$CONTAINERD_CONFIG_FILEPATH.bak"
-SOCI_RELEASE_VERSION="0.7.0"
-SOCI_TAR_CHECKSUM="8766cdd479272dcc86299e70a0f7a9343f940c98285c1491bb3c3cdc05b26f47"
-SOCI_TAR_FILENAME="soci-snapshotter-$SOCI_RELEASE_VERSION-linux-amd64.tar.gz"
-SOCI_CONFIG_DIR="/etc/soci-snapshotter-grpc/"
-SOCI_CONFIG_FILEPATH="$SOCI_CONFIG_DIR/config.toml"
-SOCI_SYSTEMD_SERVICE_FILEPATH="/etc/systemd/system/soci-snapshotter.service"
-cp $CONTAINERD_CONFIG_FILEPATH $BACKUP_CONTAINERD_CONFIG_FILEPATH
-
-setup_soci() {( set -ex
-
-yum install fuse -y
-modprobe fuse
-
-curl \
---silent \
---show-error \
---retry 3 \
---retry-delay 1 \
--L https://github.com/awslabs/soci-snapshotter/releases/download/v${SOCI_RELEASE_VERSION}/soci-snapshotter-${SOCI_RELEASE_VERSION}-linux-amd64.tar.gz \
--o $SOCI_TAR_FILENAME
-echo "$SOCI_TAR_CHECKSUM $SOCI_TAR_FILENAME" | sha256sum -c
-
-tar -C /usr/local/bin -xvf $SOCI_TAR_FILENAME soci soci-snapshotter-grpc
-rm $SOCI_TAR_FILENAME
-
-cat > $SOCI_SYSTEMD_SERVICE_FILEPATH << EOF
-[Unit]
-Description=soci snapshotter containerd plugin
-Documentation=https://github.com/awslabs/soci-snapshotter
-After=network.target
-Before=containerd.service
-
-[Service]
-Type=notify
-ExecStart=/usr/local/bin/soci-snapshotter-grpc
-Restart=always
-RestartSec=5
-
-[Install]
-WantedBy=multi-user.target
-EOF
-
-mkdir -p $SOCI_CONFIG_DIR
-
-cat > $SOCI_CONFIG_FILEPATH << EOF
+# Set environment variables
+ARCH=$(uname -m | sed s/aarch64/arm64/ | sed s/x86_64/amd64/)
+version="0.9.0"
+ARCHIVE=soci-snapshotter-$version-linux-$ARCH.tar.gz
+
+pushd /tmp
+# Download, verify, and install the soci-snapshotter
+curl --silent --location --fail --output $ARCHIVE https://github.com/awslabs/soci-snapshotter/releases/download/v$version/$ARCHIVE
+curl --silent --location --fail --output $ARCHIVE.sha256sum https://github.com/awslabs/soci-snapshotter/releases/download/v$version/$ARCHIVE.sha256sum
```

Comment:
I am curious. Shouldn't this be hardcoded in the script?

If an attacker can compromise `https://github.com/awslabs/soci-snapshotter/releases/download/v$version/$ARCHIVE`, then they can also compromise `https://github.com/awslabs/soci-snapshotter/releases/download/v$version/$ARCHIVE.sha256sum`.

Reply:
Ideally yes, MITM attack can always be done.

Reply:
we discussed on this and we will download the SHA from the github only. The reason to do so is to maintain multiple SHA for different ARCH which is not feasible.

Reply:
> maintain multiple SHA for different ARCH which is not feasible
There are only two arch 🙃

I don't think SOCI even ships binaries for other than amd64 and arm64